### PR TITLE
roam-research: add aarch64-darwin and x86_64-darwin support

### PR DESCRIPTION
**Pull request by jakuzure**
## Description of changes

Adding aarch64-darwin and x86_64-darwin support for Roam Research. It's just doing a .dgm unpack, since the source is closed.

My changes were inspired by this PR for Lens: https://github.com/NixOS/nixpkgs/pull/205348

@dbalan, if you're unable/unwilling to also be listed as a maintainer for Darwin, I could add myself instead 

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/office/roam-research/linux.nix [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+  ];
+in stdenv.mkDerivation rec {
+  pname = "roam-research";
+  version = "0.0.18";
+
+  src = fetchurl {
+    url =
+      "https://roam-electron-deploy.s3.us-east-2.amazonaws.com/${pname}_${version}_amd64.deb";
+    sha256 = "sha256-veDWBFZbODsdaO1UdfuC4w6oGCkeVBe+fqKn5XVHKDQ=";
+  };
```
Comment (Artturin) [inline pkgs/applications/office/roam-research/linux.nix:49]: Common not being used here

Comment (jakuzure) [reply pkgs/applications/office/roam-research/linux.nix:49]: Oops, nice catch, will fix it shortly, thanks!

File: pkgs/applications/office/roam-research/linux.nix [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+  nativeBuildInputs = [ dpkg ];
+
+  unpackPhase = ''
+    mkdir pkg
+    dpkg-deb -x $src pkg
+    sourceRoot=pkg
+  '';
```
Comment (Artturin) [inline pkgs/applications/office/roam-research/linux.nix:57]: dpkg now has a hook to unpack sources so this likely isn't needed anymore

Comment (jakuzure) [reply pkgs/applications/office/roam-research/linux.nix]: seems like you're right, I removed that section

File: pkgs/applications/office/roam-research/darwin.nix [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+in
+stdenv.mkDerivation rec {
+  inherit (common) pname version;
+  src = common.sources.${system} or (throw "Source for ${pname} is not available for ${system}");
+
+  appName = "Roam Research";
+
+  sourceRoot = "${appName}.app";
+
+  buildInputs = [ undmg ];
```
Comment (Artturin) [inline pkgs/applications/office/roam-research/darwin.nix:14]: This is run during the build, so it should be in `nativeBuildInputs`

File: pkgs/applications/office/roam-research/darwin.nix [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  installPhase = ''
+    mkdir -p "$out/Applications/${appName}.app"
+    cp -R . "$out/Applications/${appName}.app"
+  '';
```
Comment (Artturin) [inline pkgs/applications/office/roam-research/darwin.nix:18]:
```suggestion
  installPhase = ''
    runHook preInstall
    
    mkdir -p "$out/Applications/${appName}.app"
    cp -R . "$out/Applications/${appName}.app"
    
    runHook postInstall
  '';
```

Comment (Artturin) [reply pkgs/applications/office/roam-research/darwin.nix:18]: adding the binary to `$out/bin` allows for `nix run` to work on darwin

**Comment by Artturin** [comment]
Please change the `roam-research = callPackage ...` in `all-packages.nix` to 

```nix
  # Using import because otherwise `.override` would override the `default.nix` arguments.
  roam-research = import ../applications/office/roam-research {
    inherit stdenv callPackage;
  };
```

```nix
nix-repl> lib.functionArgs roam-research.override
{ callPackage = false; stdenv = false; }
```

```nix
nix-repl> lib.functionArgs roam-research.override
{ alsa-lib = false; atk = false; cairo = false; cups = false; dbus = false; dpkg = false; expat = false; fetchurl = false; gdk-pixbuf = false; glib = false; gtk3 = false; lib = false; libX11 = false; libXScrnSaver = false; libXcomposite = false; libXcursor = false; libXdamage = false; libXext = false; libXfixes = false; libXi = false; libXrandr = false; libXrender = false; libXtst = false; libdrm = false; libpulseaudio = false; libxcb = false; libxkbcommon = false; libxshmfence = false; mesa = false; nspr = false; nss = false; pango = false; stdenv = false; udev = false; }
```

**Comment by Artturin** [comment]
After my suggestions the pr should be ok, but we should wait a bit for the maintainer

**Comment by jakuzure** [comment]
Added your remaining suggestions, thanks for the detailed review!

File: pkgs/applications/office/roam-research/darwin.nix [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+  common = import ./common.nix { inherit fetchurl; };
+  inherit (stdenv.hostPlatform) system;
+in
+stdenv.mkDerivation rec {
+  inherit (common) pname version;
+  src = common.sources.${system} or (throw "Source for ${pname} is not available for ${system}");
+
+  appName = "Roam Research";
+
+  sourceRoot = "${appName}.app";
```
Comment (wegank) [inline pkgs/applications/office/roam-research/darwin.nix:12]:
```suggestion
  sourceRoot = ".";
```

File: pkgs/applications/office/roam-research/darwin.nix [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+  nativeBuildInputs = [ undmg ];
+  installPhase = ''
+    runHook preInstall
+
+    mkdir -p "$out/Applications/${appName}.app"
+    cp -R . "$out/Applications/${appName}.app"
```
Comment (wegank) [inline pkgs/applications/office/roam-research/darwin.nix:19]:
```suggestion
    mkdir -p "$out/Applications"
    cp -R *.app "$out/Applications"
```

File: pkgs/applications/office/roam-research/darwin.nix [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+    mkdir -p $out/bin
+    ln -s "$out/Applications/${appName}.app/Contents/MacOS/${appName}" "$out/bin/${appName}"
+    runHook postInstall
+  '';
+
+  meta = with lib; {
+    description = "A note-taking tool for networked thought.";
```
Comment (wegank) [inline pkgs/applications/office/roam-research/darwin.nix:27]:
```suggestion
    description = "A note-taking tool for networked thought";
```

File: pkgs/applications/office/roam-research/linux.nix [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+    mv usr/* "$out/"
+
+    substituteInPlace $out/share/applications/roam-research.desktop \
+      --replace "/opt/Roam Research/roam-research" "roam-research"
+  '';
+
+  dontPatchELF = true;
+  meta = with lib; {
+    description = "A note-taking tool for networked thought.";
```
Comment (wegank) [inline pkgs/applications/office/roam-research/linux.nix:64]:
```suggestion
    description = "A note-taking tool for networked thought";
```

File: pkgs/applications/office/roam-research/common.nix [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+let
+  pname = "roam-research";
+  version = "0.0.18";
+in
+{
+  inherit pname version;
+  sources = {
+    x86_64-darwin = fetchurl {
+      sha256 = "sha256-jyFNH3qrgrsftExL/b2t8bY3W3fYVz+Gp11AuaIMxbg=";
+      url = "https://roam-electron-deploy.s3.us-east-2.amazonaws.com/Roam+Research-${version}.dmg";
```
Comment (wegank) [inline pkgs/applications/office/roam-research/common.nix:11]:
```suggestion
      url = "https://roam-electron-deploy.s3.us-east-2.amazonaws.com/Roam+Research-${version}.dmg";
      hash = "sha256-jyFNH3qrgrsftExL/b2t8bY3W3fYVz+Gp11AuaIMxbg=";
```

File: pkgs/top-level/all-packages.nix [hunk 10]
```diff
@@ -12593,7 +12593,10 @@ with pkgs;
 
   rnv = callPackage ../tools/text/xml/rnv { };
 
-  roam-research = callPackage ../applications/office/roam-research { };
+  # Using import because otherwise `.override` would override the `default.nix` arguments.
+  roam-research = import ../applications/office/roam-research {
+      inherit stdenv callPackage;
+    };
```
Comment (wegank) [inline pkgs/top-level/all-packages.nix:12599]:
```suggestion
  roam-research = callPackage ../applications/office/roam-research { };
```

See `pkgs/applications/audio/spotify/default.nix` for an example.

Comment (jakuzure) [reply pkgs/top-level/all-packages.nix:12599]: So I'd have to change default.nix to this?

```{ stdenv, callPackage, ... }@args:
let
  extraArgs = removeAttrs args [ "callPackage" ];
in
if stdenv.isDarwin then
  callPackage ./darwin.nix (extraArgs // { })
else
  callPackage ./linux.nix (extraArgs // { })
```

Comment (wegank) [reply pkgs/top-level/all-packages.nix:12599]: Yes, please.

**Pull request merged**